File: 8.4.2023/ex1.py
from abc import ABC, abstractmethod

class Person(ABC):
    @abstractmethod
    def input(self):
        pass

    @abstractmethod
    def display(self):
        pass


class Student(Person):
    def __init__(self, name, age, native_place, student_id):
        self.name = name
        self.age = age
        self.native_place = native_place
        self.student_id = student_id

    def input(self):
        self.name = input("Enter name: ")
        self.age = int(input("Enter age: "))
        self.native_place = input("Enter native place: ")
        self.student_id = input("Enter student ID: ")

    def display(self):
        print(f"Name: {self.name}")
        print(f"Age: {self.age}")
        print(f"Native Place: {self.native_place}")
        print(f"Student ID: {self.student_id}")


class ManagerStudent(Student):
    def __init__(self, n):
        self.n = n
        self.students = []

    def input(self):
        for i in range(self.n):
            student = Student(None, None, None, None)
            print(f"Enter data for student {i + 1}:")
            student.input()
            self.students.append(student)

    def display(self):
        for student in self.students:
            student.display()

    def sort_by_name(self):
        self.students.sort(key=lambda student: student.name)

    def search_by_name(self, name):
        for student in self.students:
            if student.name == name:
                return student
        return None


class UsingManagerStudent:
    @staticmethod
    def menu():
        print("1. Input data.")
        print("2. Display data.")
        print("3. Sort by name.")
        print("4. Search by name.")
        print("5. Exit.")
        return int(input("Enter your choice: "))

def main():
    manager_student = ManagerStudent(int(input("Enter number of students: ")))

    while True:
        choice = UsingManagerStudent.menu()

        if choice == 1:
            manager_student.input()
        elif choice == 2:
            manager_student.display()
        elif choice == 3:
            manager_student.sort_by_name()
        elif choice == 4:
            name = input("Enter the name to search: ")
            student = manager_student.search_by_name(name)
            if student:
                student.display()
            else:
                print("Student not found.")
        elif choice == 5:
            print("Exiting...")
            break
        else:
            print("Invalid choice. Please try again.")

if __name__ == "__main__":
    main()

File: 8.4.2023/ex1.java
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Scanner;

abstract class Person {
    abstract void input();
    abstract void display();
}

class Student extends Person {
    String name;
    int age;
    String nativePlace;
    String studentId;

    public Student(String name, int age, String nativePlace, String studentId) {
        this.name = name;
        this.age = age;
        this.nativePlace = nativePlace;
        this.studentId = studentId;
    }

    void input() {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter name: ");
        name = scanner.nextLine();
        System.out.print("Enter age: ");
        age = scanner.nextInt();
        scanner.nextLine(); // Consume newline left-over
        System.out.print("Enter native place: ");
        nativePlace = scanner.nextLine();
        System.out.print("Enter student ID: ");
        studentId = scanner.nextLine();
    }

    void display() {
        System.out.println("Name: " + name);
        System.out.println("Age: " + age);
        System.out.println("Native Place: " + nativePlace);
        System.out.println("Student ID: " + studentId);
    }
}

class ManagerStudent {
    int n;
    List<Student> students;

    public ManagerStudent(int n) {
        this.n = n;
        students = new ArrayList<>();
    }

    void input() {
        for (int i = 0; i < n; i++) {
            System.out.println("Enter data for student " + (i + 1) + ":");
            Student student = new Student(null, 0, null, null);
            student.input();
            students.add(student);
        }
    }

    void display() {
        for (Student student : students) {
            student.display();
        }
    }

    void sortByName() {
        students.sort(Comparator.comparing(student -> student.name));
    }

    Student searchByName(String name) {
        for (Student student : students) {
            if (student.name.equals(name)) {
                return student;
            }
        }
        return null;
    }
}

public class ex1 {
    static int menu() {
        System.out.println("1. Input data.");
        System.out.println("2. Display data.");
        System.out.println("3. Sort by name.");
        System.out.println("4. Search by name.");
        System.out.println("5. Exit.");
        Scanner scanner = new Scanner(System.in);
        return scanner.nextInt();
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter number of students: ");
        int n = scanner.nextInt();
        ManagerStudent managerStudent = new ManagerStudent(n);

        while (true) {
            int choice = menu();
            scanner.nextLine(); // Consume newline left-over

            if (choice == 1) {
                managerStudent.input();
            } else if (choice == 2) {
                managerStudent.display();
            } else if (choice == 3) {
                managerStudent.sortByName();
            } else if (choice == 4) {
                System.out.print("Enter the name to search: ");
                String name = scanner.nextLine();
                Student student = managerStudent.searchByName(name);
                if (student != null) {
                    student.display();
                } else {
                    System.out.println("Student not found.");
                }
            } else if (choice == 5) {
                System.out.println("Exiting...");
                break;
            } else {
                System.out.println("Invalid choice. Please try again.");
            }
        }
    }
}
